---
title: "Which green tea compounds in plasma associate with multivariate composition of the microbiome?"
author: "John Sterrett"
date: '2024-03-06'
output:
  html_document: default
  pdf_document: default
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load(tidyverse,
               data.table,
               foreach,
               parallel,
               doSNOW,
               vegan,
               usedist,
               selbal,
               phyloseq)

'%ni%' <- Negate('%in%')

setwd("/Users/johnsterrett/Research-Projects/Reisdorph/GT-micro-metabo/analysis")

```

# Load data

```{r load_data, echo=F}
# read in genus level table
uni <- qiime2R::read_qza("../microbiome/core-metrics/unweighted_unifrac_distance_matrix.qza")$data 


# read in metadata
metadata <- fread("../metadata/mapping.tsv", sep = "\t")
rownames(metadata) <- metadata$`MetaboSampleID`

# read in metabolites
aqueous <- fread("../metabolome/raw_aqueous.tsv",
                 sep="\t", header = T) %>% as.data.frame()
aqueous <- aqueous %>% remove_rownames() %>% column_to_rownames(var="V1")

aqueous <- aqueous[rownames(aqueous) %ni% c("Compound Name", "Retention Time",
                                          "Measured Retention Time", "Product Mass",
                                         "QualifierMass", "CASNumber",
                                         "CompoundType", "ISTDCompoundName", 
                                         "Mass Column"),]
aqueous <- mutate_all(aqueous, function(x) as.numeric(as.character(x)))


# read in lipids
lipids <- fread("../metabolome/raw_lipid.tsv",
                sep="\t",
                header = T)

rownames(lipids) <- lipids$V1
lipids <- lipids %>% remove_rownames() %>% column_to_rownames(var="V1")

lipids <- lipids[rownames(lipids) %ni% c("Compound Name", "Retention Time",
                                          "Measured Retention Time", "Product Mass",
                                         "QualifierMass", "CASNumber",
                                         "CompoundType", "ISTDCompoundName", 
                                         "Mass Column"),]
lipids <- mutate_all(lipids, function(x) as.numeric(as.character(x)))

# map genus sample ids to metabolomics sample ids
new_sample_ids <- c()
for (x in labels(uni)){
    sampleid <- metadata[metadata$`#SampleID`==x, "MetaboSampleID"][[1,1]]
    new_sample_ids <- c(new_sample_ids, sampleid)
}

uni <- dist_setNames(uni, new_sample_ids)

# sort dataframes
uni <- uni %>% dist_subset(order(as.numeric(labels(uni))))
metadata <- metadata[order(as.numeric(rownames(metadata))),,drop=FALSE]
aqueous <- aqueous[order(as.numeric(rownames(aqueous))),,drop=FALSE]
lipids <- lipids[order(as.numeric(rownames(lipids))),,drop=FALSE]
 
# may remove the germ-free mice because with them we mostly get signal from the 
# GF vs control differences
non_controls <- metadata$PID!="Control"
controls <- !non_controls
```

```{r, aggregate pseudoreplicates}
# Create a column with unique control IDs because they ??aren't?? pseudoreplicates 
# like the humanized microbiomes are
metadata[controls, "PID_unique_controls"] <- paste0(
    unlist(metadata[controls, "PID"]), 
    1:sum(controls))
metadata[non_controls, "PID_unique_controls"] <- metadata[non_controls, "PID"]

# Aggregate control IDs by experiment cohort
# There are 3 controls per cohort, 
# and there's a bit of a cohort effect in the controls
metadata[controls, "PID_grouped_controls"] <- paste0(
    unlist(metadata[controls, "PID"]), 
    unlist(metadata[controls, "Group"]))
metadata[non_controls, "PID_grouped_controls"] <- metadata[non_controls, "PID"]


no_pr_metadata <- metadata %>% 
    group_by(PID_grouped_controls) %>%
    summarize(PID=first(PID))

# Aggregate the pseudoreplicated microbiome distances based on centroid distances
no_pr_uni <- dist_multi_centroids(uni, as.factor(metadata$PID_grouped_controls))

# Aggregate GT compounds as the mean of each humanized microbiome ID
no_pr_lipids <- lipids %>% 
    group_by(metadata$PID_grouped_controls) %>%
    summarise_all(mean)

no_pr_aqueous <- aqueous %>% 
    group_by(metadata$PID_grouped_controls) %>%
    summarise_all(mean)

# combine lipids and aqueous
gt_compounds <- merge(no_pr_aqueous, no_pr_lipids, 
                      by="metadata$PID_grouped_controls")
rownames(gt_compounds) <- gt_compounds$`metadata$PID_grouped_controls`
gt_compounds$`metadata$PID_grouped_controls` <- NULL

# rename columns
colnames(gt_compounds) <- gsub(pattern="_[0-9.]+$", replacement="", 
                           colnames(gt_compounds), perl=T)
colnames(gt_compounds) <- gsub(pattern=" - [0-9.]+$", replacement="", 
                           colnames(gt_compounds), perl=T)
colnames(gt_compounds) <- gsub(pattern="\\*\\*", replacement="", 
                           colnames(gt_compounds))


gt_compounds_cleannames <- janitor::clean_names(gt_compounds)
```

```{r run adonis}

run_adonis_for_compound <- function(distance_matrix, data,
                                    compound, 
                                    covariates_formula=NULL,
                                    permutations=9999){
    if(is.null(covariates_formula)){
        this_formula <- as.formula(paste0("distance_matrix",
                                          "~ ", compound))
    } else {
        this_formula <- as.formula(paste0("distance_matrix",
                                          "~ ", covariates_formula, 
                                          "+", compound))
    }
    res <- adonis2(this_formula, data, permutations=permutations)
    res
}

gf_index <- no_pr_metadata$PID=="Control"
adonis_results <- c()
lfc_results <- c()
t_results <- c()

for(compound in colnames(gt_compounds_cleannames)){
    this_result <- run_adonis_for_compound(no_pr_uni, 
                                           gt_compounds_cleannames, 
                                           compound=compound,
                                           permutations=99999)
    adonis_results <- c(adonis_results,
                        this_result[compound, "Pr(>F)"])
    gf_mean <- mean(gt_compounds_cleannames[gf_index, compound])
    hu_mean <- mean(gt_compounds_cleannames[!gf_index, compound])
    lfc <- log2(hu_mean/gf_mean)
    lfc_results <- c(lfc_results, lfc)
    
    t_p <- t.test(gt_compounds_cleannames[gf_index, compound],
                  gt_compounds_cleannames[!gf_index, compound])$p.value
    t_results <- c(t_results, t_p)
    
}

adonis_df <- data.frame(compound=make.unique(colnames(gt_compounds)),
                        p=adonis_results,
                        padj=p.adjust(adonis_results, method="fdr"),
                        lfc_hu_vs_gf=lfc_results,
                        t_test_p=t_results,
                        t_test_padj=p.adjust(t_results, method="fdr"))

adonis_df[adonis_df$p < 0.01,] %>%
    knitr::kable()

adonis_df %>% 
    filter(p < 0.1) %>%
    arrange(p) %>%
    mutate(p=round(p, digits=3),
           padj=round(padj, digits=3),
           lfc_hu_vs_gf=round(lfc_hu_vs_gf, digits=2)) %>%
    dplyr::select(-c(t_test_p, t_test_padj)) %>%
    write.csv("adonis_lfc_results_p_0.1.csv")

p <- EnhancedVolcano::EnhancedVolcano(toptable=adonis_df, x="lfc_hu_vs_gf", y="p",
                                 lab=gsub(pattern="\\.[0-9]$", replacement="", 
                                          adonis_df$compound, perl=T), 
                                 ylim=1e-4, pCutoff=1e-2,
                                 xlim=c(-7, 7), xlab=expression(Log[2](HU/GF)),
                                 title=NULL, subtitle=NULL,
                                 drawConnectors=TRUE, arrowheads=F)
p

ggsave("../figures/adonis_lfc_no_pr_volcano.pdf", p, height=8, width=8)
ggsave("../figures/adonis_lfc_no_pr_volcano.png", p, height=8, width=8)
```


```{r run_faith_lm}
faith <- qiime2R::read_qza("../microbiome/core-metrics/faith_pd_vector.qza")$data

row.names(faith) <- match(row.names(faith), metadata$`#SampleID`)
faith <- data.frame(faith_pd=faith[order(as.numeric(row.names(faith))),])

regress_faith_clustered_errors <- function(data,
                                    compound, 
                                    covariates_formula=NULL,
                                    cluster_col=NULL){
    if(is.null(covariates_formula)){
        this_formula <- as.formula(paste0("faith_pd",
                                          "~ ", compound))
    } else {
        this_formula <- as.formula(paste0("faith_pd",
                                          "~ ", covariates_formula, 
                                          "+", compound))
    }
    res <- fixest::feols(this_formula, data, cluster=cluster_col)
    res
}

# Removes each sample from the dataframe, reruns feols for dataframe with msising
# estimate, and returns TRUE if no samples have a dffits > threshold
# Will return FALSE if any sample has dffits > threshold
# There has to be a way to do this based on hat values, but I can't
# find it for clustered standard errors...
check_dffits_clustered_faith_feols <- function(estimate,
                                         df_threshold,
                                         data,
                                         compound, 
                                         covariates_formula=NULL,
                                         cluster_col=NULL){
    for (i in seq_len(nrow(data))) {
        missing_row_df <- data[-i, , drop=FALSE]
        missing_row_cluster_col <- cluster_col[-i]
        minus_one_result <- regress_faith_clustered_errors(missing_row_df,
                                           compound=compound,
                                           cluster_col=missing_row_cluster_col)
        minus_one_estimate <- minus_one_result$coeftable[compound, "Estimate"]
        if(abs(estimate - minus_one_estimate) > df_threshold){
            return(FALSE)
        }
    }
    
    return(TRUE)
}


# merge all the data from all samples
all_gt <- merge(lipids, aqueous, by="row.names")
rownames(all_gt) <- all_gt$Row.names
all_gt$Row.names <- NULL
all_faith_gt <- merge(faith, all_gt, by="row.names")
rownames(all_faith_gt) <- all_faith_gt$Row.names
all_faith_gt$Row.names <- NULL


# Scale all compounds re
all_faith_gt <- all_faith_gt %>% 
    mutate_at(scale, .vars=vars(-c(faith_pd)))

colnames(all_faith_gt) <- gsub(pattern="_[0-9.]+$", replacement="", 
                           colnames(all_faith_gt), perl=T)
colnames(all_faith_gt) <- gsub(pattern=" - [0-9.]+$", replacement="", 
                           colnames(all_faith_gt), perl=T)
colnames(all_faith_gt) <- gsub(pattern="\\*\\*", replacement="", 
                           colnames(all_faith_gt))

# order it like metadata
all_faith_gt <- all_faith_gt[rownames(metadata),]

all_faith_gt_cleannames <- janitor::clean_names(all_faith_gt)

p <- c()
effect <- c()

for(compound in colnames(dplyr::select(all_faith_gt_cleannames, -c(faith_pd)))){
    this_result <- regress_faith_clustered_errors(all_faith_gt_cleannames,
                                           compound=compound,
                                           cluster_col=metadata$PID_grouped_controls)
    
    dffits_passes <- check_dffits_clustered_faith_feols(
                              estimate=this_result$coeftable[compound, "Estimate"],
                              df_threshold=1, 
                              data=all_faith_gt_cleannames,
                              compound=compound,
                              cluster_col=metadata$PID)
    if(dffits_passes){
        p <- c(p, this_result$coeftable[compound, "Pr(>|t|)"])
        effect <- c(effect, this_result$coeftable[compound, "Estimate"])
    } else {
        p <- c(p, NA)
        effect <- c(effect, NA)
    }
}

# Need to use make.unique because there are a few duplicated compound names
feols_df <- data.frame(compound=make.unique(colnames(all_faith_gt))[2:ncol(all_faith_gt)],
                        p=p,
                        padj=p.adjust(p, method="fdr"),
                        effect=effect)

feols_df %>%
    filter(padj < 0.05) %>%
    knitr::kable()

feols_df %>%
    filter(is.na(p)) %>%
    knitr::kable(caption="Compounds with dffits > 1")

# remove the pattern of (6Z,9Z)
z_pattern <- "\\(\\d+Z(?:,\\d+Z)*\\)"
feols_df$compound_clean <- str_replace_all(feols_df$compound, z_pattern, "")


p <- EnhancedVolcano::EnhancedVolcano(toptable=feols_df, x="effect", y="p",
                                 lab=gsub(pattern="\\.[0-9]$", replacement="", 
                                          feols_df$compound_clean, perl=T),
                                 ylim=1e-8, pCutoff=1e-2,
                                 xlim=c(-3, 3), xlab=expression(beta),
                                 title=NULL, subtitle=NULL,
                                 drawConnectors=TRUE, arrowheads=F)
p

ggsave("../figures/feols_faith_compounds_volcano.pdf", p, height=8, width=8)
ggsave("../figures/feols_faith_compounds_volcano.png", p, height=8, width=8)


feols_df %>% 
    filter(p < 0.1) %>%
    arrange(p) %>%
    mutate(p=round(p, digits=3),
           padj=round(padj, digits=3),
           effect=round(effect, digits=2)) %>%
    write.csv("feols_faith_results_p_0.1.csv")

compounds_to_plot_faith <- feols_df %>%
    filter(padj < 0.05) %>%
    dplyr::select(compound) %>%
    unlist()
names(compounds_to_plot_faith) <- NULL

all_faith_gt_uniquenames <- all_faith_gt
colnames(all_faith_gt_uniquenames) <- make.unique(colnames(all_faith_gt_uniquenames))


faith_pd_plotdata <- all_faith_gt_uniquenames %>%
    dplyr::select(compounds_to_plot_faith,
                    faith_pd)
faith_pd_plotdata <-merge(faith_pd_plotdata, 
                          metadata$PID!="Control", by="row.names")
rownames(faith_pd_plotdata) <- faith_pd_plotdata$Row.names
faith_pd_plotdata$Row.names <- NULL
faith_pd_plotdata <- dplyr::rename(faith_pd_plotdata, "Humanized"="y")

faith_pd_plotdata <- faith_pd_plotdata %>%
    pivot_longer(cols=-c(faith_pd,
                         Humanized)) 
faith_pd_plotdata$effect <- feols_df[match(faith_pd_plotdata$name, 
                                         feols_df$compound),
                                   "effect"]
faith_pd_plotdata <- faith_pd_plotdata %>%
    mutate(name=gsub(pattern="\\.[0-9]$", replacement="", 
                    name, perl=T))

faith_pd_plotdata$name <- str_replace_all(faith_pd_plotdata$name, z_pattern, "")

faith_pd_plotdata$name <- faith_pd_plotdata$name %>% 
    reorder(faith_pd_plotdata$effect, FUN=first)



p <- ggplot(faith_pd_plotdata, 
            mapping=aes(x=faith_pd, y=value[,1],
                        #shape=Humanized
                        )) +
    geom_point(size=3) +
    geom_smooth(method="lm") +
    theme_bw() +
    facet_wrap(~name, scales="free", ncol=4,
               labeller=labeller(name=label_wrap_gen(width=50, 
                                                     multi_line=TRUE))) +
    labs(x="Faith's phylogenetic diversity", y="Compound abundance (scaled)")
p

ggsave("../figures/feols_faith_compounds_scatter.pdf", p, height=8, width=16)
ggsave("../figures/feols_faith_compounds_scatter.png", p, height=8, width=16)
```



```{r gf_vs_hu}
regress_group_clustered_errors <- function(data,
                                    compound, 
                                    group,
                                    covariates_formula=NULL,
                                    cluster_col=NULL){
    if(is.null(covariates_formula)){
        this_formula <- as.formula(paste0(compound,
                                          "~ ", group))
    } else {
        this_formula <- as.formula(paste0(compound,
                                          "~ ", covariates_formula, 
                                          "+", group))
    }
    res <- fixest::feols(this_formula, data, cluster=cluster_col)
    res
}

# Removes each sample from the dataframe, reruns feols for dataframe with msising
# estimate, and returns TRUE if no samples have a dffits > threshold
# Will return FALSE if any sample has dffits > threshold
# There has to be a way to do this based on hat values, but I can't
# find it for clustered standard errors...
check_dffits_clustered_group_feols <- function(estimate,
                                         df_threshold,
                                         data,
                                         compound, 
                                         group,
                                         covariates_formula=NULL,
                                         cluster_col=NULL){
    for (i in seq_len(nrow(data))) {
        missing_row_df <- data[-i, , drop=FALSE]
        missing_row_cluster_col <- cluster_col[-i]
        minus_one_result <- regress_group_clustered_errors(missing_row_df,
                                           compound=compound,
                                           group=group,
                                           covariates_formula=covariates_formula,
                                           cluster_col=missing_row_cluster_col)
        
        minus_one_estimate <- minus_one_result$coeftable[group, "Estimate"]
        if(abs(estimate - minus_one_estimate) > df_threshold){
            return(FALSE)
        }
    }
    
    return(TRUE)
}

# Sort it like metadata
all_gt_hu <- all_gt[rownames(metadata),]

# Scale data
all_gt_hu <- all_gt_hu %>% 
    mutate_all(scale)

all_gt_hu$humanized <- metadata$PID=="Control"

colnames(all_gt_hu) <- gsub(pattern="_[0-9.]+$", replacement="", 
                           colnames(all_gt_hu), perl=T)
colnames(all_gt_hu) <- gsub(pattern=" - [0-9.]+$", replacement="", 
                           colnames(all_gt_hu), perl=T)
colnames(all_gt_hu) <- gsub(pattern="\\*\\*", replacement="", 
                           colnames(all_gt_hu))

all_gt_cleannames <- janitor::clean_names(all_gt_hu)

# Code this as numeric for easy access of coefficient
all_gt_cleannames <- all_gt_cleannames %>%
    mutate(humanized=as.numeric(humanized))

p_hu <- c()
effect_hu <- c()

for(compound in colnames(dplyr::select(all_gt_cleannames, -c(humanized)))){
    this_result <- regress_group_clustered_errors(all_gt_cleannames,
                                           compound=compound,
                                           group="humanized",
                                           cluster_col=metadata$PID_grouped_controls)
    
    dffits_passes <- check_dffits_clustered_group_feols(
                              estimate=this_result$coeftable["humanized", "Estimate"],
                              df_threshold=1, 
                              data=all_gt_cleannames,
                              compound=compound,
                              group="humanized",
                              cluster_col=metadata$PID_grouped_controls)
    if(dffits_passes){
        p_hu <- c(p_hu, this_result$coeftable["humanized", "Pr(>|t|)"])
        effect_hu <- c(effect_hu, this_result$coeftable["humanized", "Estimate"])
    } else {
        p_hu <- c(p_hu, NA)
        effect_hu <- c(effect_hu, NA)
    }
}

# Need to use make.unique because there are a few duplicated compound names
feols_hu_df <- data.frame(compound=make.unique(colnames(all_gt_hu))[2:ncol(all_gt_hu)],
                        p=p_hu,
                        padj=p.adjust(p_hu, method="fdr"),
                        effect=effect_hu)

feols_hu_df %>%
    filter(padj < 0.1) %>%
    knitr::kable()

feols_hu_df %>%
    filter(is.na(p)) %>%
    knitr::kable(caption="Compounds with dffits > 1")


feols_hu_df[feols_hu_df$p < 0.1, "compound"]

feols_hu_df %>% 
    filter(p < 0.1) %>%
    arrange(p) %>%
    mutate(p=round(p, digits=3),
           padj=round(padj, digits=3),
           effect=round(effect, digits=2)) %>%
    write.csv("feols_hu_results_p_0.1.csv")

```



# Selbal
```{r fig.height=6, fig.width=14, warning=FALSE}
pseq <- qiime2R::qza_to_phyloseq("../microbiome/core-metrics/rarefied_table.qza", 
                        taxonomy="../microbiome/taxonomy_GT.qza")

pseq <- pseq %>% tax_glom(taxrank="Genus", NArm=FALSE)
genus_table <- pseq %>% otu_table()

if(identical(rownames(otu_table(pseq)), taxa_names(pseq))){
    new_names <- paste0(
        tax_table(pseq)[taxa_names(pseq), "Family"],
        "_",
        tax_table(pseq)[taxa_names(pseq), "Genus"]   
    )
    new_names[new_names=="NA_NA"] <- 
        paste0(new_names[new_names=="NA_NA"],
               1:sum(new_names=="NA_NA"))
    rownames(genus_table) <- new_names
}

if(length(unique(colSums(genus_table)))>1){
    stop("ColSums aren't identical. Watch out.")
}

genus_table <- genus_table %>%
    t() %>%
    as.data.frame() %>%
    group_by(metadata$PID_grouped_controls) %>%
    summarise_all(mean) %>%
    dplyr::select(-c(`metadata$PID_grouped_controls`))


get_balance_vals <- function(selbal_res){
    mod <- selbal_res$glm
    mod$model$V1
}

get_global_balance_r2 <- function(selbal_res){
    mod <- selbal_res$glm
    cor(mod$fitted, mod$y)^2
}

all_selbals <- data.frame(matrix(nrow=0, ncol=6))
colnames(all_selbals) <- c("compound", "compound_abundance", 
                           "balance_values",
                           "balance_num", "balance_den",
                           "r2")
    
gt_compounds_uniquenames <- gt_compounds
colnames(gt_compounds_uniquenames) <- make.unique(colnames(gt_compounds_uniquenames))
for (compound in adonis_df[adonis_df$p < 0.01, "compound"]){
    print(compound)
    
    selbal_res <- selbal.cv(x=genus_table, y=gt_compounds_uniquenames[,compound],
                        n.fold=5, n.iter=10)

    global_balance <- selbal_res$global.balance
    
    balance_values <- get_balance_vals(selbal_res)
    r2 <- get_global_balance_r2(selbal_res)
    
    res <- data.frame(
               compound=rep(compound, length(balance_values)),
               compound_abundance=gt_compounds_uniquenames[,compound],
               
               balance_values=balance_values,
               
               balance_num=global_balance[global_balance$Group=="NUM", "Taxa"],
               balance_den=global_balance[global_balance$Group=="DEN", "Taxa"],
               
               r2=rep(r2, length(balance_values))
               )
    all_selbals <- rbind(all_selbals, res)
}

all_selbals$compound <- str_replace_all(all_selbals$compound, z_pattern, "")

all_selbals$facet_name <- paste0(gsub(pattern="\\.[0-9]$", replacement="", 
                                      all_selbals$compound, perl=T), 
                                 ", ", expression(R^2), " = ",
                                 round(all_selbals$r2, 2))
all_selbals$facet_name <- all_selbals$facet_name %>% 
    reorder(1-all_selbals$r2, FUN=first)

all_selbals$balance_den[duplicated(all_selbals$facet_name)] <- NA
all_selbals$balance_num[duplicated(all_selbals$facet_name)] <- NA
all_selbals$balance_den <- all_selbals$balance_den %>% 
    gsub(pattern="\\[", replacement="") %>%
    gsub(pattern="\\]", replacement="")
all_selbals$balance_num <- all_selbals$balance_num %>% 
    gsub(pattern="\\[", replacement="") %>%
    gsub(pattern="\\]", replacement="")


p <- ggplot(all_selbals, mapping=aes(x=balance_values, y=compound_abundance)) +
    geom_point(size=3) +
    geom_smooth(method="lm") +
    geom_text(mapping = aes(label = ifelse(is.na(balance_num), "", paste0("  Num: ", balance_num))),
            x=-Inf, y=Inf, hjust=0, vjust=1.2, size=3, na.rm=T) +
    geom_text(mapping = aes(label = ifelse(is.na(balance_den), "", paste0("  Den: ", balance_den))),
            x=-Inf, y=Inf, hjust=0, vjust=2.3, size=3, na.rm=T) +
    theme_bw() +
    facet_wrap(~facet_name, scales="free", ncol=4) +
    labs(x="Balance", y="Compound abundance")

p
ggsave("../figures/selbal_compounds.pdf", plot=p, height=6, width = 14)
ggsave("../figures/selbal_compounds.png", plot=p, height=6, width = 14)
```

# Procrustes randomization test
```{r protest_setup}
uni_pcoa <- ape::pcoa(no_pr_uni)
uni_pcoa_coords <- uni_pcoa$vectors

aqueous_bc <- no_pr_aqueous %>% 
    dplyr::select(-c(`metadata$PID_grouped_controls`)) %>%
    ecodist::bcdist()
aqueous_bc_pcoa <- ape::pcoa(aqueous_bc)
aqueous_bc_pcoa_coords <- aqueous_bc_pcoa$vectors

lipids_bc <- no_pr_lipids %>% 
    dplyr::select(-c(`metadata$PID_grouped_controls`)) %>%
    ecodist::bcdist()
lipids_bc_pcoa <- ape::pcoa(lipids_bc)
lipids_bc_pcoa_coords <- lipids_bc_pcoa$vectors
```

```{r run_protest}
protest(uni_pcoa_coords[,1:10], aqueous_bc_pcoa_coords[,1:10], permutations=9999)
protest(uni_pcoa_coords[,1:10], lipids_bc_pcoa_coords[,1:10], permutations=9999)


mantel(no_pr_uni, aqueous_bc, permutations=9999)
mantel(no_pr_uni, lipids_bc, permutations=9999)
```

```{r adonis_by_group}
adonis2(no_pr_uni ~ no_pr_metadata$PID=="Control")

adonis2(aqueous_bc ~ no_pr_metadata$PID=="Control")
adonis2(lipids_bc ~ no_pr_metadata$PID=="Control")
```


```{r adonis_by_cohort}
no_pr_group_metadata <- metadata %>%
    group_by(PID_grouped_controls) %>%
    dplyr::summarise(Group=first(Group))

adonis2(no_pr_uni ~ no_pr_group_metadata$Group)

adonis2(aqueous_bc ~ no_pr_group_metadata$Group)
adonis2(lipids_bc ~ no_pr_group_metadata$Group)
```


```{r faith_by_group}
if(sum(rownames(faith)==rownames(metadata)) != nrow(metadata)){
    stop("faith and metadata rownames don't match up")
}
feols_group_df <- merge(faith, metadata, by="row.names")
feols_group_df$humanized <- feols_group_df$PID != "Control"

fixest::feols(faith_pd ~ humanized, data=feols_group_df, cluster=feols_group_df$PID)

lmerTest::lmer(faith_pd ~ humanized + (1|PID), data=feols_group_df) %>% summary()
```

