---
title: "Taxa-bar"
author: "John Sterrett"
date: "2023-03-07"
output:
  pdf_document: default
  html_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

pacman::p_load(tidyverse,
               data.table,
               ggplot2,
               phyloseq,
               patchwork,
               microshades,
               cowplot,
               magrittr)
#setwd("analysis")
```

# Introduction
## Document focus & Microshades issue
In this document, I demonstrate an issue with the way relative abundances are calculated in Microshades. That is, when taxa have not been assigned taxonomy at the taxonomic level used for microshading (such as genus), they are dropped from all consideration, and relative abundances are calculated **after** these taxa are removed. However, relative abundances calculated without taxa unassigned at the genus (or other) level are not representative of the actual relative abundances. 

## Dataset
Our dataset shows an extremely clear example of this. This dataset consists of 16S sequencing data from 20 humanized mice (microbiomes were humanized in pairs, meaning 2 mice received the same human microbiome) + 9 "germ-free" mice. There was some contamination in the "germ-free" mice, but they remained with very simple (low diversity) microbiome composition, dominated by just a few ASVs from the phylum Firmicutes.

## How I noticed this issue
One of these ASVs that was high relative abundance in the "germ-free" controls belonged to the family peptostreptococcaceae. In a few "germ-free" samples, it makes up >75% of the microbiome. However, it did not have a genus-level taxonomic classification. Therefore, when I make a microshades plot with microshading at the genus level, it is suddenly >90% *Turicibacter*, which is **not** in the peptostreptococcaceae family. This is a massive change in the presentation of the data, and you can see it by clicking through the tabbed microshades plots below.

This issue also leads to changes in the non-microshaded colors. For example, see the samples humanized with microbiome #74. When microshading at the genus level, the firmicutes:bacteroidetes ratio (purple vs blue) appears to be half of what it actually is. **I'd like to reiterate that a lack of a genus-level taxonomic classification is changing the visualization of known classifications at the phylum level.**

## Solution
In this document, I show 2 potential solutions. One is much more naive; it simply renames any NA to be "unclassified". You can see as a proof of concept that the phylum level coloring (e.g. purple vs blue) no longer gets affected by unclassified taxa. The second solution is the one I prefer. It loops through the taxonomy table and labels any unassigned levels as "unassigned_<parent>", where parent is the lowest level of taxonomy assigned above the unassigned level. In the case of our ASV that is in the peptostreptococcaceae family but has an unassigned genus, it labeled it "Unassigned_Peptostreptococcaceae", which shows up at the proper relative abundance in the resulting microshades plot.

# Load data
```{r}
pseq <- qiime2R::qza_to_phyloseq(feature="../microbiome/filtered_table_GT.qza",
                                 taxonomy="../microbiome/taxonomy_GT.qza",
                                 metadata="../metadata/mapping_nocomment.tsv")

taxonomy.table <- pseq %>% tax_table() %>% as.data.frame()
```

# Microshades {.tabset}
```{r}
make_microshades_phylum <- function(pseq, subgroup){
    # prep the microshades colors
    mdf_prep <- prep_mdf(pseq, subgroup_level=subgroup)
    # sort the phylum names
    phylum_table <- tax_glom(pseq, taxrank="Phylum", ) %>% otu_table()
    phyla.otunames <- rownames(phylum_table)
    
    phylums <- taxonomy.table[phyla.otunames,"Phylum"]
    
    sorted_phylums <- phylums[order(rowSums(phylum_table), decreasing=T)]
    # create the colors object
    color_objs_GP <- create_color_dfs(mdf_prep, selected_groups=sorted_phylums[5:1], 
                                      cvd = TRUE, subgroup_level=subgroup)
    # Extract
    mdf_GP <- color_objs_GP$mdf
    cdf_GP <- color_objs_GP$cdf
    # create a custom legend
    GP_legend <-custom_legend(mdf_GP, cdf_GP, 
                              legend_key_size=unit(0.4, "cm"),
                              legend_text_size=10, subgroup_level=subgroup)
    
    # plot
    plot <- plot_microshades(mdf_GP, cdf_GP)
    plot_1 <- plot + scale_y_continuous(labels = scales::percent, expand = expansion(0)) +
      theme(legend.position = "none")  +
      #theme(axis.text.x = element_blank()) +
      facet_grid(~PID, scales="free_x", space="free_x"
                 )
    
    multi <- plot_grid(plot_1, GP_legend,  rel_widths = c(1, .25))
    multi
}

```

**Summary: Click through the tabs below to see the changes in phylum level relative abundances based on the sub-level for microshades.**

In the Phylum->Genus plot below, it appears that the genus Turicibacter makes up >90% in our controls. However, we know this isn't the case if we look at the Phylum-Family plot (Turicibacteraceae is not high relative abundance). I double-check this in the section below, and Turicibacter is at no more than 23% in any sample.

While this may be an extreme example, it could have considerable ramifications here and in other cases. And in some cases, these might not be caught. For example, in the Phylum-Genus plot, the Fimicutes:Bacteroidetes ratio appears to be half of what it actually is for the samples from humanized microbiome #74. That could lead to a lot of confusion when comparing statistical results to the taxa bar plot!

## Phylum -> Order
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(pseq, "Order")

#ggsave("../figures/taxa-bar-phyum-genus-microshades.pdf",plot=multi,height=10,width=10)
```

## Phylum -> Family
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(pseq, "Family")
```


## Phylum -> Genus
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(pseq, "Genus")
```

# Check relative abundances
```{r}
temp_phylum_table <- tax_glom(pseq, taxrank="Phylum") %>% otu_table()
phyla.otunames <- rownames(temp_phylum_table)
rownames(temp_phylum_table) <- taxonomy.table[phyla.otunames,"Phylum"]

taxonomy.table <- taxonomy.table %>% as.data.frame()
Turici.otunames <- taxonomy.table %>%
    filter(Genus=="Turicibacter") %>% 
    rownames()

otu.table <- pseq %>% otu_table() %>% as.data.frame()
rel.otu.table <- ( t(otu.table)/colSums(otu.table) ) %>% t()
print(paste("Max relative abundance of Turicibacter in any sample",
            max(rel.otu.table[Turici.otunames,]))
      )

```

# Redo after making these unclassified taxa say "unclassified" {.tabset}

If you click through the tabs, you can see that changing the microshades sub-level no longer changes the phylum relative abundances. However, this isn't the most elegant solution.
```{r}

rep_na_with_unclassified <- function(x){
    x[is.na(x)] <- "unclassified"
    return(x)
}

no.na.taxonomy.table <- taxonomy.table %>% 
    mutate_all(rep_na_with_unclassified) %>% 
    as.data.frame() %>%
    as.matrix()

no.na.pseq <- pseq
tax_table(no.na.pseq) <- tax_table(no.na.taxonomy.table)
```

## Phylum -> Order
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(no.na.pseq, "Order")
```

## Phylum -> Family
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(no.na.pseq, "Family")
```

## Phylum -> Genus
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(no.na.pseq, "Genus")
```

# Replace NA with informative unclassified {.tabset}
```{r}
rep_na_with_unclassified_above <- function(pseq){
    # get the tax table
    tax.tab <- pseq %>% tax_table() %>% as.data.frame()
    # get the number of taxonomic levels
    cols <- ncol(tax.tab)
    
    # for each asv
    for (asv in rownames(tax.tab)){
        # subset
        curr.asv <- tax.tab[asv,]
        # initialize parent unfound
        parent.found <- F; i <- cols
        # This starts at the lowest taxonomic level, 
        # then moves up the taxonomy (the if condition) until it finds non-NA
        # If it hits a non-NA, it replaces all lower levels of taxonomy with 
        # "Unclassified <parent>" and sets parent.found to TRUE, so the loop
        # will exit. 
        # Then, curr.asv replaces the old row in the taxonomic table.
        while(!parent.found){
            if(is.na(curr.asv[i])){
                i <- i-1
            }
            else{
                parent.found <- T
                cols.to.update <- (i+1):(cols+1)
                curr.asv[cols.to.update[cols.to.update <= cols]] <- paste0("Unnasigned_", curr.asv[i])
            }
        }
        tax.tab[asv,] <- curr.asv
    }
    # returns the tax table matrix
    return (as.matrix(tax.tab))
    
}

new.tax.table <- rep_na_with_unclassified_above(pseq)

pseq.inf.unclass <- pseq
tax_table(pseq.inf.unclass) <- new.tax.table
```

## Phylum -> Order
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(pseq.inf.unclass, "Order")
```

## Phylum -> Family
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(pseq.inf.unclass, "Family")
```

## Phylum -> Genus
```{r, warning=F, echo=F, fig.height=8}
make_microshades_phylum(pseq.inf.unclass, "Genus")
ggsave("../figures/phylum_genus_taxa_bar.pdf", height=8, width=10.5)
```

# Old vs new tax table
Can compare old vs new taxonomy
```{r}
taxonomy.table %>% 
    head(n=10) %>% 
    knitr::kable(caption="Old Taxonomy")

new.tax.table %>% 
    as.data.frame() %>%
    head(n=10) %>% 
    knitr::kable(caption="New Taxonomy")
```

# Characterize top ASVs in GF mice
```{r}
metadata <- sample_data(pseq)
gf <- metadata$PID=="Control"
top.gf.asvs.order <- order(rowSums(rel.otu.table[,gf]), decreasing=T)

top.gf.asvs <- head(rel.otu.table[top.gf.asvs.order,]) %>%
    rownames()
# get mean relative abund of each asv
top.gf.asvs.abund <- head(rel.otu.table[top.gf.asvs.order,]) %>%
    rowSums()
top.gf.asvs.abund <- top.gf.asvs.abund/sum(gf) # dividing sum by num of samples
top.gf.asvs.abund <- top.gf.asvs.abund*100

top.gf.asvs.tax <- taxonomy.table[top.gf.asvs,]

# create table
tax.abund.table <- merge(top.gf.asvs.tax[,c("Class", "Order", 
                                            "Family", "Genus")],
                         top.gf.asvs.abund, 
                         by="row.names")
tax.abund.table <- tax.abund.table %>% 
    rename("Relative Abundance"="y",
           "ASV ID"="Row.names")
tax.abund.table <- tax.abund.table[order(tax.abund.table$`Relative Abundance`,
                                   decreasing=T),
                                   ]
tax.abund.table %>% 
    knitr::kable(digits=1)

tax.abund.table$`Relative Abundance` <-tax.abund.table$`Relative Abundance`%>% 
    round(digits=1)

tax.abund.table %>% write_csv("tax_abund_gf_asvs.csv")

tax.abund.table
```

## Blast the rep seqs for these ASVs
```{r blasting away}

rep.seqs <- qiime2R::read_qza("../microbiome/rep-seqs_GT.qza")$data 

# grab only our seqs
rep.seqs.of.interest <- rep.seqs %>% as.data.frame()
rep.seqs.of.interest <- rep.seqs.of.interest[tax.abund.table$`ASV ID`,]
rep.seqs.of.interest <- Biostrings::DNAStringSet(rep.seqs.of.interest)

# get blast 16s db
if(!dir.exists("16S_ribosomal_RNA")){
    download.file("https://ftp.ncbi.nlm.nih.gov/blast/db/16S_ribosomal_RNA.tar.gz",
                  "16S_ribosomal_RNA.tar.gz", mode='wb')

    untar("16S_ribosomal_RNA.tar.gz", exdir="16SMicrobialDB")    
}

# read in blast db
bl <- rBLAST::blast(db="./16SMicrobialDB/16S_ribosomal_RNA")

# start blasting
cl <- predict(bl, rep.seqs.of.interest, BLAST_args = "-perc_identity 99", 
              custom_format = "qseqid pident length Mismatches stitle")
cl


```